File: fixtures/coding/batch-ingestion-go/filter.go
package pipeline

// Record filtering — sign your implementation here.
//
// All function signatures are fixed. Fill the function bodies.

// FilterRule is a single filter criterion.
type FilterRule struct {
	Field    string
	Operator string // eq, neq, contains, gt, lt
	Value    any
}

// MatchesFilterRule checks whether a record satisfies a single filter rule.
//
// Supported operators:
//   - eq / neq: equality comparison on any field
//   - contains: value (string) in tags OR value in string payload fields
//   - gt / lt: timestamp comparison on TS only, rule value is an RFC3339 string
//
// Field may be: "record_type", "ts", "payload:<key>", "tags".
// Return true if the rule is satisfied.
func MatchesFilterRule(record Record, rule FilterRule) bool {
	return false
}

// ApplyFilters returns only records that satisfy ALL rules (AND logic).
//
// Preserve original order.
// Empty rules list returns all records unchanged.
func ApplyFilters(records []Record, rules []FilterRule) []Record {
	return nil
}

File: fixtures/coding/batch-ingestion/pipeline/filter.py
"""Record filtering — sign your implementation here.

All function signatures are fixed. Fill the function bodies.
"""

from dataclasses import dataclass
from typing import Any

from pipeline.types import Record


@dataclass(frozen=True)
class FilterRule:
    """A single filter criterion."""

    field: str
    operator: str  # eq, neq, contains, gt, lt
    value: Any


def matches_filter_rule(record: Record, rule: FilterRule) -> bool:
    """Check whether a record satisfies a single filter rule.

    Supported operators:
      - eq / neq:   equality comparison on any field
      - contains:   value (str) in tags OR value (str) in
                    str(payload[key]) for payload fields
      - gt / lt:    numeric comparison on 'ts' only (as unix ts)

    'field' may be: "record_type", "ts", "payload:<key>", "tags"
    Return True if the rule is satisfied.
    """
    ...


def apply_filters(records: list[Record], rules: list[FilterRule]) -> list[Record]:
    """Return only records that satisfy ALL rules (AND logic).

    Preserve original order.
    Empty rules list → return all records unchanged.
    """
    ...
